runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -359,16 +543,82 @@ public void finishBundle() {
         emitResults();
       } catch (Exception e) {
         throw new RuntimeException("Failed to finish remote bundle", e);
+      } finally {
+        remoteBundle = null;
+      }
+      if (bundleFinishedCallback != null) {
+        bundleFinishedCallback.run();
+        bundleFinishedCallback = null;
       }
     }
 
+    boolean isBundleInProgress() {
+      return remoteBundle != null;
+    }
+
+    void setBundleFinishedCallback(Runnable callback) {
+      this.bundleFinishedCallback = callback;
+    }
+
     private void emitResults() {
       KV<String, OutputT> result;
       while ((result = outputQueue.poll()) != null) {
-        outputManager.output(outputMap.get(result.getKey()), (WindowedValue) result.getValue());
+        final String inputCollectionId = result.getKey();
+        TupleTag<?> tag = outputMap.get(inputCollectionId);
+        WindowedValue windowedValue =
+            Preconditions.checkNotNull(
+                (WindowedValue) result.getValue(),
+                "Received a null value from the SDK harness for %s",
+                inputCollectionId);
+        if (tag != null) {
+          // process regular elements
+          outputManager.output(tag, windowedValue);
+        } else {
+          // process timer elements
+          // TODO This is ugly. There should be an easier way to retrieve the
+          String timerPCollectionId =
+              inputCollectionId.substring(0, inputCollectionId.length() - ".out:0".length());
```

Comment:
@lukecwik I couldn't figure out how to retrieve the correct name for the collection of the timer input. I believe this needs to be fixed.

Reply:
This looks ugly indeed.. Is this going to go away with the changes @robertwb mentioned last week?

Reply:
There is #6986 but this just prepares the Java code for handling the Python representation.

Reply:
Found a way, for example:

```java
stageBundleFactory
.getProcessBundleDescriptor()
// yes this is required twice
.getProcessBundleDescriptor()
.getPcollectionsMap()
.get("ParDo(Anonymous)2/ParMultiDo(Anonymous).foo.out:0")
.getUniqueName()
```
=> `"ParDo(Anonymous)2/ParMultiDo(Anonymous).foo"`